python/src/iceberg/expression/literals.py

```diff
@@ -0,0 +1,502 @@
+    def _(self, type_var: DoubleType) -> "DoubleLiteral":
+        return DoubleLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DoubleLiteral(Literal[float]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DoubleType)
+    def _(self, type_var: DoubleType) -> "DoubleLiteral":
+        return self
+
+    @to.register(FloatType)
+    def _(self, type_var: FloatType) -> Union[AboveMax, BelowMin, FloatLiteral]:
+        if FloatType.max < self.value:
+            return AboveMax()
+        elif FloatType.min > self.value:
+            return BelowMin()
+        return FloatLiteral(self.value)
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> "DecimalLiteral":
+        if type_var.scale == 0:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal("1."), rounding=ROUND_HALF_UP))
+        else:
+            return DecimalLiteral(Decimal(self.value).quantize(Decimal((0, (1,), -type_var.scale)), rounding=ROUND_HALF_UP))
+
+
+class DateLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return self
+
+
+class TimeLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimeType)
+    def _(self, type_var: TimeType) -> "TimeLiteral":
+        return self
+
+
+class TimestampLiteral(Literal[int]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(TimestampType)
+    def _(self, type_var: TimestampType) -> "TimestampLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.fromtimestamp(self.value / 1000000) - EPOCH).days)
+
+
+class DecimalLiteral(Literal[Decimal]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(DecimalType)
+    def _(self, type_var: DecimalType) -> Optional["DecimalLiteral"]:
+        if type_var.scale == abs(self.value.as_tuple().exponent):
+            return self
+        return None
+
+
+class StringLiteral(Literal[str]):
+    @singledispatchmethod
+    def to(self, type_var):
+        return None
+
+    @to.register(StringType)
+    def _(self, type_var: StringType) -> "StringLiteral":
+        return self
+
+    @to.register(DateType)
+    def _(self, type_var: DateType) -> "DateLiteral":
+        return DateLiteral((datetime.datetime.strptime(self.value, "%Y-%m-%d") - EPOCH).days)
```

Comment:
I think it will be more reliable to use `date` rather than `datetime`:

```python
from datetime.datetime import date
EPOCH_DATE = date.fromisoformat("1970-01-01") # OR date(1970, 1, 1)
...
return DateLiteral((date.fromisoformat(self.value) - EPOCH_DATE).days)
```

Reply:
I did some testing for `date.fromisoformat` and it correctly validates the number of digits.